threads.go

```diff
@@ -0,0 +1,384 @@
+func (t *threads) Delete(ctx context.Context, id thread.ID) error {
+	panic("implement me")
+}
+
+// getPrivKey returns the host's private key.
+func (t *threads) getPrivKey() ic.PrivKey {
+	return t.host.Peerstore().PrivKey(t.host.ID())
+}
+
+// createLog call util.CreateLog.
+func (t *threads) createLog() (info thread.LogInfo, err error) {
+	return util.CreateLog(t.host.ID())
+}
+
+// getOrCreateLog returns the log with the given thread and log id
+// If no log exists, a new one is created under the given thread.
+func (t *threads) getOrCreateLog(id thread.ID, lid peer.ID) (info thread.LogInfo, err error) {
+	info = t.LogInfo(id, lid)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+// getOwnLoad returns the log owned by the host under the given thread.
+func (t *threads) getOwnLog(id thread.ID) (info thread.LogInfo) {
+	for _, lid := range t.LogsWithKeys(id) {
+		if t.PrivKey(id, lid) != nil {
+			info = t.LogInfo(id, lid)
+			return
+		}
+	}
+	return
+}
+
+// getOrCreateOwnLoad returns the log owned by the host under the given thread.
+// If no log exists, a new one is created under the given thread.
+func (t *threads) getOrCreateOwnLog(id thread.ID) (info thread.LogInfo, err error) {
+	info = t.getOwnLog(id)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+// createRecord creates a new record with the given body as a new event body.
+func (t *threads) createRecord(ctx context.Context, body format.Node, lg thread.LogInfo, settings *tserv.AddSettings) (thread.Record, error) {
+	if settings.Key == nil {
+		var key []byte
+		logKey := t.ReadKey(settings.Thread, settings.KeyLog)
+		if logKey != nil {
+			key = logKey
+		} else {
+			key = lg.ReadKey
+		}
+		var err error
+		settings.Key, err = crypto.ParseEncryptionKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	event, err := cbor.NewEvent(ctx, t.dagService, body, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	prev := cid.Undef
+	if len(lg.Heads) != 0 {
+		prev = lg.Heads[0]
+	}
+	fk, err := crypto.ParseEncryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+	rec, err := cbor.NewRecord(ctx, t.dagService, event, prev, lg.PrivKey, fk)
+	if err != nil {
+		return nil, err
+	}
+
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+
+	return rec, nil
+}
+
+// pullLocal returns local records from the given thread that are ahead of offset but not farther then limit.
+func (t *threads) pullLocal(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, limit int) ([]thread.Record, error) {
+	lg := t.LogInfo(id, lid)
+	if lg.PubKey == nil {
+		return nil, fmt.Errorf("could not find log")
+	}
+	fk, err := crypto.ParseDecryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var recs []thread.Record
+	if limit == 0 {
+		return recs, nil
+	}
+	cursor := lg.Heads[0]
```

Comment:
I remember that `len(lg.Heads)>0` is a future-idea, but maybe for the moment assert that `==1`?